packages/service/ni_measurement_plugin_sdk_service/_internal/parameter/serialization_descriptors.py

```diff
@@ -0,0 +1,177 @@
+        )
+    else:
+        enum_type_name = _get_enum_type(parameter_metadata).__name__
+
+    if enum_type_name == "int" or enum_type_name is None:
+        field_descriptor.type_name = ProtobufColor.DESCRIPTOR.full_name
+    elif enum_type_name not in [enum_type.name for enum_type in file_descriptor.enum_type]:
+        enum_descriptor = file_descriptor.enum_type.add()
+        enum_descriptor.name = enum_type_name
+        for name, number in enum_dict.items():
+            enum_value_descriptor = enum_descriptor.value.add()
+            enum_value_descriptor.name = name
+            enum_value_descriptor.number = number
+        field_descriptor.type_name = enum_descriptor.name
+    else:
+        field_descriptor.type_name = enum_type_name
+
+
+def _get_enum_type(parameter_metadata: ParameterMetadata) -> type:
+    if parameter_metadata.repeated and len(parameter_metadata.default_value) > 0:
+        return type(parameter_metadata.default_value[0])
+    else:
+        return type(parameter_metadata.default_value)
+
+
+def _create_field(
+    message_proto: Any, metadata: ParameterMetadata, index: int
+) -> FieldDescriptorProto:
+    """Implement a field in 'message_proto'.
+
+    Args:
+        message_proto (message_type): A message instance in a file descriptor proto.
+
+        metadata (ParameterMetadata): Metadata of 'param'.
+
+        index (int): 'param' index in parameter_values
+
+    Returns:
+        FieldDescriptorProto: field_descriptor of 'param'.
+    """
+    field_descriptor = message_proto.field.add()
+    field_descriptor.number = index
+    field_descriptor.name = metadata.sanitized_display_name()
+    field_descriptor.type = metadata.type
+
+    if metadata.repeated:
+        field_descriptor.label = FieldDescriptorProto.LABEL_REPEATED
+        field_descriptor.options.packed = True
+    else:
+        field_descriptor.label = FieldDescriptorProto.LABEL_OPTIONAL
+
+    if metadata.type == FieldDescriptorProto.TYPE_MESSAGE:
+        field_descriptor.type_name = metadata.message_type
+    return field_descriptor
+
+
+def create_file_descriptor(
+    service_name: str,
+    output_metadata: List[ParameterMetadata],
+    input_metadata: List[ParameterMetadata],
+    pool: descriptor_pool.DescriptorPool,
+) -> None:
+    """Creates two message types in one file descriptor proto.
+
+    Args:
+        service_class_name (str): Unique service name.
+
+        output_metadata (List[ParameterMetadata]): Metadata of output parameters.
+
+        input_metadata (List[ParameterMetadata]): Metadata of input parameters.
+
+        pool (DescriptorPool): Descriptor pool holding file descriptors and enum classes.
+
+    Returns:
+        None: Only creates a file and two message descriptors.
+    """
+    service_name = "".join(char for char in service_name if char.isalpha())
+    try:
+        pool.FindFileByName(service_name)
+    except KeyError:
```

Comment:
Good, this handles the case where the file descriptor is already defined.

It would also make sense for this to compare the previously registered file descriptor to the configuration/output metadata and error if they don't match.

Reply:
Is there a concern that these will change? As far as I could think, they'll get created upon start of the service and put into the pool. After that, it won't change unless the Python process also restarts in which case the pool will be empty again.

Would this check just be checking that we added it correctly during service start or is there some other concern about it changing after service start?

Reply:
Yes, the concern is that a test or an advanced user will start/stop services multiple times with slightly different config/output parameters. I don't think it matters for basic usage like in the example services.

If FileDescriptor messages are comparable, then it would be easy to do this check: generate the FileDescriptor again and compare the old and new FileDescriptors. 

If it's not that easy, I think it's ok not to bother.

Reply:
Me and Joel couldn't find a easy way to compare these because both file descriptors return different types from each other and it also compares the pointers, not the contents in them.

Reply:
Thanks, I'll drop this issue.